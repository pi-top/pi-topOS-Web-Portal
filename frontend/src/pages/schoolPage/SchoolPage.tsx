import React, { useState } from "react";
import { Line as ProgressBar } from "rc-progress";

import Layout from "../../components/layout/Layout";
import Upload from "../../components/upload/Upload";
import updateImage from "../../assets/images/upgrade-page.png";
import styles from "./SchoolPage.module.css";

const getErrorMessage = (error: string) => {
  return `There was a problem with your bundle; please try again (${error}).`;
};

export type Props = {
  onBackButtonClick?: () => void;
};

export default ({
  onBackButtonClick,
}: Props) => {
  const [uploadSucceeded, setUploadSucceeded] = useState(false);
  const [isUploading, setIsUploading] = useState(false);
  const [error, setError] = useState("");
  const [progress, setProgress] = useState(0);

  const getPrompt = () => {
    if (uploadSucceeded) {
      return <>Bundle uploaded <span className="green">successfully</span>!</>;
    }
    return <>Let's upload your <span className="green">setup bundle</span>!</>;
  }

  const getExplanation = () => {
    if (uploadSucceeded) {
      return <>
        <p>Please follow the instructions in the pi-top[4] miniscreen to complete the setup.</p>
        <p>Once the process is complete, sign up on <a className={styles.link} href="https://further.pi-top.com">further.pi-top.com</a> to start teaching with your device.</p>
      </>
    }
    return <>
      <p>If you don't have a setup bundle, download one <a className={styles.link} href="https://further.pi-top.com/setup-devices">here</a>.</p>
      <p>When configuring multiple devices, it may be faster to put this file on a USB and plug it into each pi-top[4].</p>
    </>
  }

  const onProgress = (progressEvent: ProgressEvent) => {
    const { loaded, total } = progressEvent;
    let percent = Math.floor((loaded * 100) / total);
    setProgress(percent);
  };

  return (
    <Layout
      banner={{
        src: updateImage,
        alt: "school-banner",
      }}
      prompt={getPrompt()}
      explanation={getExplanation()}
      nextButton={{}}
      showNext={false}
      showBack={onBackButtonClick !== undefined }
      backButton={{
        onClick: onBackButtonClick,
        disabled: isUploading,
      }}
      className={styles.root}
    >
      {!isUploading && !uploadSucceeded && (
        <Upload
          userInstruction="Click to upload your bundle or drag and drop it here"
          filenameRegex={/^pi-top-usb-setup\.tar\.gz$/}
          onUploadStart={() => {
            setError("");
            setUploadSucceeded(false);
            setIsUploading(true);
          }}
          onUploadEnd={() => {
            setUploadSucceeded(true);
            setIsUploading(false);
          }}
          onUploadError={(e) => {
            setError(e);
            setIsUploading(false);
          }}
          supportedExtensions={["gz"]}
          disabled={false}
          onUploadProgress={onProgress}
        />
      )}

      { isUploading && (
        <div data-testid="progress" className={styles.progress}>
          <ProgressBar
            percent={progress}
            strokeWidth={3}
            strokeColor="#71c0b4"
          />
          <span data-testid="progress-message" className={styles.text}>
            Please wait while we upload your bundle; this could take a few minutes ...
          </span>
        </div>
      )}

      {error && (
        <span data-testid="error-message" className={styles.error}>
          {getErrorMessage(error)}
        </span>
      )}

    </Layout>
  );
};
