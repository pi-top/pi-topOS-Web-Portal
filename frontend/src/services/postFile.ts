import axios from "axios";

import apiBaseUrl from "./apiBaseUrl";

export default async function postFile(
  data: FormData,
  onUploadProgress: (progressEvent: ProgressEvent) => void = (_) => {}
) {
  await axios.post(
    `${apiBaseUrl}/upload-file`,
    data,
    {
      onUploadProgress,
      headers: {
        "Content-Type": "multipart/form-data;",
      },
    }
  );
}
