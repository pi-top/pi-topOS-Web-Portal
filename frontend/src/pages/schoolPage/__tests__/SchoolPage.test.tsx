import React from "react";
import {
  act,
  render,
  fireEvent,
  BoundFunction,
  GetByBoundAttribute,
  QueryByText,
  QueryByBoundAttribute,
  GetByText,
  waitForElement,
  getByTestId,
  wait,
} from "@testing-library/react";

import SchoolPage from "../SchoolPage";

import postFile from "../../../services/postFile";
jest.mock("../../../services/postFile");
const postFileMock = postFile as jest.Mock;

const createFile = (filename: string, size: number, type: string) => {
  const file = new File([""], filename, { type });
  Object.defineProperty(file, "size", {
    get() {
      return size;
    },
  });
  return file;
};

describe("SchoolPage", () => {
  let container: HTMLElement;
  let getByAltText: BoundFunction<GetByBoundAttribute>;
  let queryByText: BoundFunction<QueryByText>;
  let getByText: BoundFunction<GetByText>;
  let queryByTestId: BoundFunction<QueryByBoundAttribute>;

  beforeEach(() => {
    postFileMock.mockRestore();
    ({ container, getByAltText, queryByText, getByText, queryByTestId } =
      render(<SchoolPage />));
  });

  it("renders correct image", () => {
    expect(getByAltText("school-banner")).toMatchSnapshot();
  });

  it("renders prompt correctly", () => {
    const prompt = container.querySelector(".prompt");
    expect(prompt).toMatchSnapshot();
  });

  it("renders explanation", () => {
    expect(queryByTestId("layout-explanation")).toMatchSnapshot();
  });

  it("doesn't render error message", () => {
    expect(queryByTestId("error-message")).not.toBeInTheDocument();
  });

  it("doesn't render back button", () => {
    expect(queryByText("Back")).not.toBeInTheDocument();
  });

  it("doesn't render Yes button", () => {
    expect(queryByText("Yes")).not.toBeInTheDocument();
  });

  it("doesn't render skip button", () => {
    expect(queryByText("Skip")).not.toBeInTheDocument();
  });

  it("renders a drop-zone to upload files", () => {
    expect(queryByTestId("dropzone-input-wrapper")).toMatchSnapshot();
  });

  it("displays an error message if user tries to upload a file with the wrong name or extension", async () => {
    act(() => {
      const dropzone = container.querySelector("input")!;
      fireEvent.change(dropzone, {
        target: {
          files: [createFile("pi-top-usb-setup.tar", 1234, "application/tar")],
        },
      });
    });
    await waitForElement(() => queryByTestId("error-message"));
    expect(queryByTestId("error-message")).toMatchSnapshot();
  });

  it("displays an error message if user tries to upload a large file", async () => {
    act(() => {
      const dropzone = container.querySelector("input")!;
      fireEvent.change(dropzone, {
        target: {
          files: [
            createFile(
              "pi-top-usb-setup.tar.gz",
              2000000001,
              "application/tar"
            ),
          ],
        },
      });
    });
    await waitForElement(() => queryByTestId("error-message"));
    expect(queryByTestId("error-message")).toMatchSnapshot();
  });

  it("displays a progress bar while the file is being transfered", async () => {
    postFileMock.mockImplementation(
      (
        data: FormData,
        onUploadProgress: (progressEvent: ProgressEvent) => void = (_) => {}
      ) => {
        return new Promise(() => {
          onUploadProgress({ loaded: 10, total: 100 });
        });
      }
    );
    act(() => {
      const dropzone = container.querySelector("input")!;
      fireEvent.change(dropzone, {
        target: {
          files: [
            createFile("pi-top-usb-setup.tar.gz", 100, "application/tar"),
          ],
        },
      });
    });

    await waitForElement(() => queryByTestId("progress"));
    // dropzone isn't rendered anymore
    expect(container.querySelector("input")).not.toBeInTheDocument();
    // progress bar is rendered
    expect(queryByTestId("progress")).toMatchSnapshot();
  });

  it("displays a message when the transfer finishes", async () => {
    postFileMock.mockImplementation(
      (
        data: FormData,
        onUploadProgress: (progressEvent: ProgressEvent) => void = (_) => {}
      ) => {
        return new Promise((resolve) => {
          onUploadProgress({ loaded: 100, total: 100 });
          resolve("OK")
        });
      }
    );

    act(() => {
      const dropzone = container.querySelector("input")!;
      fireEvent.change(dropzone, {
        target: {
          files: [
            createFile("pi-top-usb-setup.tar.gz", 100, "application/tar"),
          ],
        },
      });
    });

    await wait();
    await wait();
    // explanation message is updated
    expect(queryByTestId("layout-explanation")).toMatchSnapshot();
  });
});
