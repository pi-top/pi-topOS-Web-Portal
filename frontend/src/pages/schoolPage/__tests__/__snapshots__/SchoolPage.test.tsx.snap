// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SchoolPage displays a message when the transfer finishes 1`] = `
<span
  class="explanation"
  data-testid="layout-explanation"
>
  <p>
    Please follow the instructions in the pi-top[4] miniscreen to complete the setup.
  </p>
  <p>
    Once the process is complete, sign up on 
    <a
      class="link"
      href="https://further.pi-top.com"
    >
      further.pi-top.com
    </a>
     to start teaching with your device.
  </p>
</span>
`;

exports[`SchoolPage displays a progress bar while the file is being transfered 1`] = `
<div
  class="progress"
  data-testid="progress"
>
  <svg
    class="rc-progress-line "
    preserveAspectRatio="none"
    viewBox="0 0 100 3"
  >
    <path
      class="rc-progress-line-trail"
      d="M 1.5,1.5
           L 98.5,1.5"
      fill-opacity="0"
      stroke="#D9D9D9"
      stroke-linecap="round"
      stroke-width="1"
    />
    <path
      class="rc-progress-line-path"
      d="M 1.5,1.5
           L 98.5,1.5"
      fill-opacity="0"
      stroke="#71c0b4"
      stroke-linecap="round"
      stroke-width="3"
      style="stroke-dasharray: 10px, 100px; stroke-dashoffset: -0px; transition: stroke-dashoffset 0.3s ease 0s, stroke-dasharray .3s ease 0s, stroke 0.3s linear; transition-duration: .3s, .3s, .3s, .06s;"
    />
  </svg>
  <span
    class="text"
    data-testid="progress-message"
  >
    Please wait while we upload your bundle; this could take a few minutes ...
  </span>
</div>
`;

exports[`SchoolPage displays an error message if user tries to upload a file with the wrong name or extension 1`] = `
<span
  class="error"
  data-testid="error-message"
>
  There was a problem with your bundle; please try again (Invalid filename).
</span>
`;

exports[`SchoolPage displays an error message if user tries to upload a large file 1`] = `
<span
  class="error"
  data-testid="error-message"
>
  There was a problem with your bundle; please try again (Maximum upload file size exceeded).
</span>
`;

exports[`SchoolPage renders a drop-zone to upload files 1`] = `
<div
  class="dropzone"
  data-testid="dropzone-input-wrapper"
  role="presentation"
  tabindex="0"
>
  <input
    class="dropzoneInput"
    data-testid="dropzone-input"
    multiple=""
    style="display: none;"
    tabindex="-1"
    type="file"
  />
  <img
    alt="add"
    class="addIcon"
    src="add.svg"
  />
  <p
    class="instruction"
    data-testid="dropzone-instruction"
  >
    Click to upload your bundle or drag and drop it here
  </p>
</div>
`;

exports[`SchoolPage renders correct image 1`] = `
<img
  alt="school-banner"
  class="image"
  draggable="false"
  src="upgrade-page.png"
  style="max-height: 100%; max-width: 100%;"
/>
`;

exports[`SchoolPage renders explanation 1`] = `
<span
  class="explanation"
  data-testid="layout-explanation"
>
  <p>
    If you don't have a setup bundle, download one 
    <a
      class="link"
      href="https://further.pi-top.com/setup-devices"
    >
      here
    </a>
    .
  </p>
  <p>
    When configuring multiple devices, it may be faster to put this file on a USB and plug it into each pi-top[4].
  </p>
</span>
`;

exports[`SchoolPage renders prompt correctly 1`] = `
<h1
  class="prompt"
>
  Let's upload your 
  <span
    class="green"
  >
    setup bundle
  </span>
  !
</h1>
`;
